pyquil/quil.py

```diff
@@ -981,3 +983,36 @@ def percolate_declares(program: Program) -> Program:
     p._defined_gates = program._defined_gates
 
     return p
+
+
+def validate_protoquil(program: Program) -> None:
+    """
+    Ensure that a program is valid ProtoQuil, otherwise raise a ValueError.
+    Protoquil allows a global RESET before any gates, and MEASUREs on each qubit after any gates
+    on that qubit. Pragmas are always allowed, and a final Halt instruction is allowed.
+
+    :param program: The Quil program to validate.
+    """
+    gates_seen = False
+    halted = False
+    measured_qubits = set()
+    for instr in program.instructions:
+        if isinstance(instr, Pragma) or isinstance(instr, Declare):
+            continue
+        elif isinstance(instr, Halt):
+            halted = True
+        elif halted:
```

Comment:
somewhat of a style choice, but you could also do this without an additional boolean `halted` flag by `enumerate`-ing the instructions and comparing the current loop index to the length of instructions when encountering a `HALT` to decide whether or not to raise an error. I think it would be a bit more readable, and would have the benefit of keeping the cases homogeneous in your big `if-elif-else` chain